[EAThread][EASTL] Added EAThread and modified EASTL

c0rp3n opened the pull request:
EABase has now also been seperated from EASTL.
EASTL now copies CommonCppFlags.cmake as it is used by EAThread.

Need to improve the patches for the includes as I am not sure on the use of `_VCPKG_ROOT_DIR` in portfiles but is it the only way I've found to fix the includes.

c0rp3n committed:
[EAThread][EASTL] Added EAThread and modified EASTL

EABase has now also been seperated from EASTL.
EASTL now copies CommonCppFlags.cmake as it is used by EAThread.

JackBoosY commented:
/azp run

JackBoosY commented:
/azp run

ras0219-msft commented on ports/eastl/fix_cmake_install.patch:
```diff
@@ -1,19 +1,28 @@
+ option(EASTL_BUILD_BENCHMARK "Enable generation of build files for benchmark" OFF)
+ option(EASTL_BUILD_TESTS "Enable generation of build files for tests" OFF)
+ 
++find_library(eabase CONFIG REQUIRED)
```
Capitalization -- this should be `EABase`.

Additionally, this probably was intended to be `find_package()`.

ras0219-msft commented on ports/eabase/fix_cmake_install.patch:
```diff
@@ -0,0 +1,35 @@
++
++    # create and install an export set for eabase target as EABase::EABase
++    install(TARGETS EABase EXPORT EABaseTargets DESTINATION ${CMAKE_INSTALL_LIBDIR})
++    install(EXPORT EABaseTargets NAMESPACE EABase:: DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/EABase)
```
This will create an `EABaseTargets.cmake`, but not `EABaseConfig.cmake` which is required for `find_package()` to operate correctly.

The easiest way to fix this would be to add `FILE EABaseConfig.cmake` to this install call.

ras0219-msft commented on ports/eastl/fix_cmake_install.patch:
```diff
@@ -1,19 +1,28 @@
-+include_directories(APPEND test/packages)
- target_include_directories(EASTL PUBLIC include)
+-target_include_directories(EASTL PUBLIC include)
++target_include_directories(EASTL PUBLIC include ${_VCPKG_ROOT_DIR}/installed/${VCPKG_TARGET_TRIPLET}/include)
```
Linking against `EABase::EABase` will cause the triplet include directory to be included -- we should strongly avoid using `${_VCPKG_ROOT_DIR}/installed/${VCPKG_TARGET_TRIPLET}` inside a patch.

Worst case, we should instead use `find_path(EABASE_INCDIR eabase.h)` or equivalent and include that folder.

ras0219-msft commented on ports/eastl/fix_cmake_install.patch:
```diff
@@ -1,19 +1,28 @@
  #-------------------------------------------------------------------------------------------
-+include_directories(APPEND test/packages)
- target_include_directories(EASTL PUBLIC include)
+-target_include_directories(EASTL PUBLIC include)
```
As above, this `include` probably needs to be replaced with the usual `$<BUILD_INTERFACE:...>` / `$<INSTALL_INTERFACE:...>` dance.

ras0219-msft commented on ports/eathread/fix_cmake_install.patch:
```diff
@@ -0,0 +1,27 @@
+ # Compiler Flags
+ #-------------------------------------------------------------------------------------------
+-set (CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH};${CMAKE_CURRENT_SOURCE_DIR}/test/packages/EASTL/scripts/CMake")
++set (CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH};${_VCPKG_ROOT_DIR}/installed/${VCPKG_TARGET_TRIPLET}/share/eastl")
```
This is unfortunate, but I don't see a better option.

ras0219-msft commented on ports/eathread/fix_cmake_install.patch:
```diff
@@ -0,0 +1,27 @@
+ # Export Include Directories
+ #-------------------------------------------------------------------------------------------
+-target_include_directories(EAThread PUBLIC include)
++target_include_directories(EAThread PUBLIC include ${_VCPKG_ROOT_DIR}/installed/${VCPKG_TARGET_TRIPLET}/include)
```
This should be handled by calling `find_package(EASTL CONFIG REQUIRED)` and linking the target.

PhoebeHui commented on ports/eastl/CONTROL:
```diff
@@ -3,3 +3,4 @@ Version: 3.14.01
 Homepage: https://github.com/electronicarts/EASTL
```
Could you also update the Version to 3.14.01-1 in this file?

PhoebeHui commented on ports/eathread/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+include(vcpkg_common_functions)
```
Could you remove 'include(vcpkg_common_functions)', this is deprecated.

c0rp3n committed:
Fix EABase

c0rp3n committed:
Fix EABASE include for EASTL

c0rp3n committed:
Improve EASTL

Still need to fix
The /debug/lib/cmake folder should be merged with /lib/cmake into /share/eastl
Found 3 error(s). Please correct the portfile:
    E:\vcpkg\ports\eastl\portfile.cmake
-- Performing post-build validation done
Error: Building package eastl:x64-windows-static failed with: POST_BUILD_CHECKS_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: eastl:x64-windows-static
  Vcpkg version: 2019.09.12-nohash

c0rp3n committed:
Base pass one Thread

PhoebeHui commented:
@c0rp3n, thanks for updates!

Could you please resolve below failures with adding 'vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/eastl)' to portfile,cmake?

The CI test failures:
The /lib/cmake folder should be merged with /debug/lib/cmake and moved to /share/eastl/cmake.
The following cmake files were found outside /share/eastl. Please place cmake files in /share/eastl.

    C:/vsts/_work/3/s/packages/eastl_x86-windows/lib/cmake/EASTL/EASTLConfig.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/lib/cmake/EASTL/EASTLConfigVersion.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/lib/cmake/EASTL/EASTLTargets-release.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/lib/cmake/EASTL/EASTLTargets.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/debug/lib/cmake/EASTL/EASTLConfig.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/debug/lib/cmake/EASTL/EASTLConfigVersion.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/debug/lib/cmake/EASTL/EASTLTargets-debug.cmake
    C:/vsts/_work/3/s/packages/eastl_x86-windows/debug/lib/cmake/EASTL/EASTLTargets.cmake

The /debug/lib/cmake folder should be merged with /lib/cmake into /share/eastl
Found 3 error(s). Please correct the portfile:

c0rp3n commented:
Thanks will do, ran out of time to resolve that will sort it today.

c0rp3n committed:
Fix config errors and add `vcpkg_fixup_cmake_targets`

c0rp3n commented:
Rebased to see if that would stop ppconsul from breaking, but it shouldnt be due to this, also x86-windows seems meant to fail atm as it isnt getting functions labeled as 64 such as `_InterlockedExchange64` which I believe is meant to come from Microsoft STL, I'm not sure whether too add guards around these functions or whether this is meant to fail.

JackBoosY commented:
@c0rp3n  Ignore it, ppconsul is fixing in #9752.

JackBoosY commented on ports/eathread/portfile.cmake:
```diff
@@ -0,0 +1,32 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
I found that the arm / uwp build failed. If this port does not support them, please add `vcpkg_fail_port_install(ON_TARGET "UWP" ON_ARCH "arm" "arm64")` here and add the following code to _VCPKG_PATH/scripts/ci.baseline.txt_:
```
eathread:arm64-windows=fail
eathread:arm-uwp=fail
eathread:x64-uwp=fail
```

c0rp3n committed:
Add required changes for fails

c0rp3n committed:
x86-windows should also fail

c0rp3n committed:
Revert adding `vcpkg_fail_port_install(ON_TARGET "Windows" ...)`

c0rp3n commented:
Should be fixed now, added the correct fails to `ci.bastline.txt` and added the port fail for UWP as requested.

JackBoosY commented:
@c0rp3n Please remove `eastl:arm64-windows=fail` from _VCPKG_PATH/scripts/ci.baseline.txt_, thanks!
Except this, LGTM.

c0rp3n committed:
Remove `eastl:arm64-windows=fail` from ci.baseline.txt

c0rp3n commented:
Done

JackBoosY commented on ports/eabase/CONTROL:
```diff
@@ -0,0 +1,5 @@
+Version: 2.09.10
+Homepage: https://github.com/electronicarts/EABase
+Description: Electronic Arts Base.
+  EABase is a small set of header files that define platform-independent data types and macros.
```
Can you merge these two lines into one?

JackBoosY commented on ports/eastl/CONTROL:
```diff
@@ -1,5 +1,6 @@
+Version: 3.16.01
 Homepage: https://github.com/electronicarts/EASTL
 Description: Electronic Arts Standard Template Library.
   It is a C++ template library of containers, algorithms, and iterators useful for runtime and tool development across multiple platforms. It is a fairly extensive and robust implementation of such a library and has an emphasis on high performance above all other considerations.
```
Can you merge these two lines into one?

c0rp3n committed:
Merge CONTROL descriptions into one line

PhoebeHui reviewed (approved):
Thanks for your updates!

ras0219-msft commented:
Thanks for this PR and addressing all our feedback!
